File: pact-jvm-consumer/src/main/java/au/com/dius/pact/consumer/dsl/PactDslWithState.java
package au.com.dius.pact.consumer.dsl;

import au.com.dius.pact.consumer.ConsumerPactBuilder;
import au.com.dius.pact.core.model.ProviderState;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class PactDslWithState {
    private final ConsumerPactBuilder consumerPactBuilder;
    List<ProviderState> state;
    String consumerName;
    String providerName;
  private final PactDslRequestWithoutPath defaultRequestValues;
  private final PactDslResponse defaultResponseValues;

  PactDslWithState(ConsumerPactBuilder consumerPactBuilder, String consumerName, String providerName,
                     ProviderState state, PactDslRequestWithoutPath defaultRequestValues,
                     PactDslResponse defaultResponseValues) {
        this(consumerPactBuilder, consumerName, providerName, defaultRequestValues, defaultResponseValues);
        this.state.add(state);
    }

    PactDslWithState(ConsumerPactBuilder consumerPactBuilder, String consumerName, String providerName,
                     PactDslRequestWithoutPath defaultRequestValues, PactDslResponse defaultResponseValues) {
      this.consumerPactBuilder = consumerPactBuilder;
      this.consumerName = consumerName;
      this.providerName = providerName;
      this.defaultRequestValues = defaultRequestValues;
      this.defaultResponseValues = defaultResponseValues;
      this.state = new ArrayList<>();
    }

    /**
     * Description of the request that is expected to be received
     *
     * @param description request description
     */
    public PactDslRequestWithoutPath uponReceiving(String description) {
        return new PactDslRequestWithoutPath(consumerPactBuilder, this, description, defaultRequestValues,
          defaultResponseValues);
    }

    /**
     * Adds another provider state to this interaction
     * @param stateDesc Description of the state
     */
    public PactDslWithState given(String stateDesc) {
      state.add(new ProviderState(stateDesc));
      return this;
    }

    /**
     * Adds another provider state to this interaction
     * @param stateDesc Description of the state
     * @param params State data parameters
     */
    public PactDslWithState given(String stateDesc, Map<String, Object> params) {
        state.add(new ProviderState(stateDesc, params));
        return this;
    }
}

File: consumer/src/main/kotlin/au/com/dius/pact/consumer/dsl/PactDslWithState.kt
package au.com.dius.pact.consumer.dsl

import au.com.dius.pact.consumer.ConsumerPactBuilder
import au.com.dius.pact.core.model.PactSpecVersion
import au.com.dius.pact.core.model.ProviderState
import au.com.dius.pact.core.support.json.JsonValue

open class PactDslWithState @JvmOverloads constructor(
  private val consumerPactBuilder: ConsumerPactBuilder,
  var consumerName: String,
  var providerName: String,
  private val defaultRequestValues: PactDslRequestWithoutPath?,
  private val defaultResponseValues: PactDslResponse?,
  val version: PactSpecVersion = PactSpecVersion.V3,
  private var additionalMetadata: MutableMap<String, JsonValue> = mutableMapOf()
) {
  @JvmField
  var state: MutableList<ProviderState> = mutableListOf()
  val comments = mutableListOf<String>()

  @Suppress("LongParameterList")
  internal constructor(
    consumerPactBuilder: ConsumerPactBuilder,
    consumerName: String,
    providerName: String,
    state: ProviderState,
    defaultRequestValues: PactDslRequestWithoutPath?,
    defaultResponseValues: PactDslResponse?,
    version: PactSpecVersion,
    additionalMetadata: MutableMap<String, JsonValue>
  ) : this(consumerPactBuilder, consumerName, providerName, defaultRequestValues, defaultResponseValues, version) {
    this.state.add(state)
    this.additionalMetadata = additionalMetadata
  }

  /**
   * Description of the request that is expected to be received
   *
   * @param description request description
   */
  fun uponReceiving(description: String): PactDslRequestWithoutPath {
    return PactDslRequestWithoutPath(consumerPactBuilder, this, description, defaultRequestValues,
      defaultResponseValues, version, additionalMetadata)
  }

  /**
   * Adds another provider state to this interaction
   * @param stateDesc Description of the state
   */
  fun given(stateDesc: String): PactDslWithState {
    state.add(ProviderState(stateDesc))
    return this
  }

  /**
   * Adds another provider state to this interaction
   * @param stateDesc Description of the state
   * @param params State data parameters
   */
  fun given(stateDesc: String, params: Map<String, Any?>): PactDslWithState {
    state.add(ProviderState(stateDesc, params))
    return this
  }

  /**
   * Adds a comment to this interaction
   */
  fun comment(comment: String): PactDslWithState {
    comments.add(comment)
    return this
  }

  /**
   * Adds additional values to the metadata section of the Pact file
   */
  fun addMetadataValue(key: String, value: String): PactDslWithState {
    additionalMetadata[key] = JsonValue.StringValue(value)
    return this
  }

  /**
   * Adds additional values to the metadata section of the Pact file
   */
  fun addMetadataValue(key: String, value: JsonValue): PactDslWithState {
    additionalMetadata[key] = value
    return this
  }
}
